In [1]:
# Importing the required libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [119]:
url = '/Users/riddhi/Documents/DS/KPMG/KPMG_VI_New_raw_data_update_final.xlsx'

transactions = pd.read_excel(url, sheet_name='Transactions')
customer_demographic = pd.read_excel(url, sheet_name='CustomerDemographic')
customer_add = pd.read_excel(url, sheet_name='CustomerAddress')
new_customer_lists = pd.read_excel(url, sheet_name='NewCustomerList')

# 1. Exploring the Transactions Dataset

In [3]:
transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [4]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

The values in the **product_first_sold_date** columns are not correct as it shows everything happening the same day at different times. 

In [11]:
#convert date columns from integer to datetime
transactions['product_first_sold_date'] = pd.to_datetime(transactions['product_first_sold_date'], unit='s')
transactions['product_first_sold_date'].head()

0   1970-01-01 11:27:25
1   1970-01-01 11:35:01
2   1970-01-01 10:06:01
3   1970-01-01 10:02:25
4   1970-01-01 11:43:46
Name: product_first_sold_date, dtype: datetime64[ns]

In [13]:
transactions.shape

(20000, 13)

In [14]:
print("Rows : ", transactions.shape[0])
print("Columns : ", transactions.shape[1])

Rows :  20000
Columns :  13


In [15]:
transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  
count   19803.000000  
mean      556.046951  
std       405.955660  
min         7.210000  
25%       215.140000  
50%       507.580000  
75%       795.100000  
max      1759.850000

### Explore missing values


In [16]:
# returns the bool value if there is any missing value in dataset
transactions.isna().values.any()

True

In [17]:
transactions.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

##### We can decide to drop missing values depending on the objective of our analysisa as there are 2000 entries in our dataset.

In [28]:
# transactions.dropna(subset=['online_order','brand','product_line','product_class','product_size','standard_cost','product_first_sold_date'])
#    __OR__
# transactions = transactions[-transactions["online_order"].isnull()]
#    __OR__
transactions.dropna(inplace=True)
transactions

transaction_id  product_id  customer_id transaction_date  online_order  \
0                   1           2         2950       2017-02-25           0.0   
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost product_first_sold_date  
0           71.49          53.62     1970-01-01 11:27:25  
1         2091.47         388.92     1970-01-01 11:35:01  
2         1793.43         248.82     1970-01-01 10:06:01  
3         1198.46         381.10     1970-01-01 10:02:25  
4         1765.30         709.48     1970-01-01 11:43:46  
...           ...            ...                     ...  
19995     2005.66        1203.40     1970-01-01 10:30:23  
19996      416.98         312.74     1970-01-01 09:52:40  
19997     1636.90          44.71     1970-01-01 11:13:30  
19998      227.88         136.73     1970-01-01 10:36:56  
19999     1775.81        1580.47     1970-01-01 10:05:34  

[19445 rows x 13 columns]

In [29]:
#to check for unique values

transactions.nunique()

transaction_id             19445
product_id                   101
customer_id                 3492
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   100
standard_cost                103
product_first_sold_date      100
dtype: int64

In [30]:
transactions.duplicated().sum()

0

In [31]:
transactions[transactions.duplicated()]  #to check for duplicate values

Empty DataFrame
Columns: [transaction_id, product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [32]:
transactions.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [43]:
transactions['online_order'].value_counts()

1.0    9739
0.0    9706
Name: online_order, dtype: int64

In [44]:
transactions['order_status'].value_counts()

Approved     19273
Cancelled      172
Name: order_status, dtype: int64

In [45]:
transactions['brand'].value_counts()

Solex             4169
WeareA2B          3245
Giant Bicycles    3244
OHM Cycles        2993
Trek Bicycles     2931
Norco Bicycles    2863
Name: brand, dtype: int64

In [46]:
transactions['product_line'].value_counts()

Standard    13920
Road         3894
Touring      1213
Mountain      418
Name: product_line, dtype: int64

In [47]:
transactions['product_class'].value_counts()

medium    13587
high       2952
low        2906
Name: product_class, dtype: int64

In [48]:
transactions['product_size'].value_counts()

medium    12767
large      3900
small      2778
Name: product_size, dtype: int64

#### All the columns appear to have consistent and correct information.


In [197]:
# will add "profit" column. Formulae = List price - Standard Cost

transactions["profit"] = transactions["list_price"] - transactions["standard_cost"]
transactions.head()


transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date   profit  
0       71.49          53.62                  41245.0    17.87  
1     2091.47         388.92                  41701.0  1702.55  
2     1793.43         248.82                  36361.0  1544.61  
3     1198.46         381.10                  36145.0   817.36  
4     1765.30         709.48                  42226.0  1055.82

In [51]:
transactions.isnull().values.any()


False

# 2. Exploring New Customer List Dataset

In [52]:
new_customer_lists.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.60      0.7500       0.9375     0.796875   
1                  11        0.64      0.6400       0.8000     0.680000   
2                   5        0.40      0.4000       0.4000     0.400000   
3                   1        1.01      1.2625       1.2625     1.262500   
4                   9        1.05      1.0500       1.3125     1.312500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [53]:
new_customer_lists.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [54]:
new_customer_lists.nunique()

first_name                              940
last_name                               961
gender                                    3
past_3_years_bike_related_purchases     100
DOB                                     958
job_title                               184
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        1
owns_car                                  2
tenure                                   23
address                                1000
postcode                                522
state                                     3
country                                   1
property_valuation                       12
Unnamed: 16                              71
Unnamed: 17                             132
Unnamed: 18                             184
Unnamed: 19                             320
Unnamed: 20                             324
Rank                                    324
Value                           

In [55]:
new_customer_lists['gender'].value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [56]:
new_customer_lists[new_customer_lists.gender == 'U']

first_name    last_name gender  past_3_years_bike_related_purchases DOB  \
59       Normy     Goodinge      U                                    5 NaT   
226      Hatti     Carletti      U                                   35 NaT   
324   Rozamond       Turtle      U                                   69 NaT   
358      Tamas      Swatman      U                                   65 NaT   
360      Tracy   Andrejevic      U                                   71 NaT   
374     Agneta      McAmish      U                                   66 NaT   
434      Gregg      Aimeric      U                                   52 NaT   
439      Johna       Bunker      U                                   93 NaT   
574    Harlene         Nono      U                                   69 NaT   
598   Gerianne       Kaysor      U                                   15 NaT   
664     Chicky      Sinclar      U                                   43 NaT   
751    Adriana  Saundercock      U                                   20 NaT   
775     Dmitri        Viant      U                                   62 NaT   
835      Porty       Hansed      U                                   88 NaT   
883      Shara     Bramhill      U                                   24 NaT   
904       Roth         Crum      U                                    0 NaT   
984    Pauline     Dallosso      U                                   82 NaT   

                        job_title job_industry_category     wealth_segment  \
59            Associate Professor                    IT      Mass Customer   
226               Legal Assistant                    IT  Affluent Customer   
324               Legal Assistant                    IT      Mass Customer   
358       Assistant Media Planner         Entertainment  Affluent Customer   
360                 Programmer II                    IT      Mass Customer   
374  Structural Analysis Engineer                    IT      Mass Customer   
434              Internal Auditor                    IT      Mass Customer   
439                Tax Accountant                    IT      Mass Customer   
574       Human Resources Manager                    IT      Mass Customer   
598               Project Manager                    IT  Affluent Customer   
664                      Operator                    IT     High Net Worth   
751                         Nurse                    IT     High Net Worth   
775                     Paralegal    Financial Services  Affluent Customer   
835               General Manager                    IT      Mass Customer   
883                           NaN                    IT  Affluent Customer   
904               Legal Assistant                    IT      Mass Customer   
984    Desktop Support Technician                    IT  Affluent Customer   

    deceased_indicator owns_car  ...  state    country  property_valuation  \
59                   N       No  ...    VIC  Australia                   5   
226                  N      Yes  ...    NSW  Australia                   9   
324                  N      Yes  ...    VIC  Australia                   3   
358                  N       No  ...    QLD  Australia                   8   
360                  N      Yes  ...    VIC  Australia                   7   
374                  N       No  ...    QLD  Australia                   6   
434                  N       No  ...    VIC  Australia                   5   
439                  N      Yes  ...    VIC  Australia                   6   
574                  N       No  ...    NSW  Australia                   7   
598                  N       No  ...    NSW  Australia                  11   
664                  N      Yes  ...    VIC  Australia                   4   
751                  N      Yes  ...    VIC  Australia                   7   
775                  N       No  ...    VIC  Australia                   1   
835                  N       No  ...    NSW  Australia                  11   
883   

In [61]:
index_name = new_customer_lists[new_customer_lists.gender == "U"].index
new_customer_lists.drop(index_name, inplace = True)
new_customer_lists[new_customer_lists.gender == 'U']

Empty DataFrame
Columns: [first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, address, postcode, state, country, property_valuation, Unnamed: 16, Unnamed: 17, Unnamed: 18, Unnamed: 19, Unnamed: 20, Rank, Value]
Index: []

[0 rows x 23 columns]

In [174]:
# Drop Unnamed Column
cols = ['Unnamed: 16','Unnamed: 17','Unnamed: 18','Unnamed: 19','Unnamed: 20','Rank','Value']
new_customer_lists = new_customer_lists.drop(cols, axis=1)

In [64]:
new_customer_lists.isna().values.any()

True

In [65]:
new_customer_lists.isna().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                              105
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
dtype: int64

In [67]:
new_customer_lists.duplicated().sum()

0

There are no duplicate values in the dataset


In [176]:
new_customer_lists['year'] = pd.DatetimeIndex(new_customer_lists['DOB']).year
new_customer_lists['month'] = pd.DatetimeIndex(new_customer_lists['DOB']).month
new_customer_lists['age'] = 2021 - new_customer_lists['year']

new_customer_lists.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0         Manufacturing      Mass Customer                  N      Yes   
1              Property      Mass Customer                  N       No   
2    Financial Services  Affluent Customer                  N       No   
3         Manufacturing  Affluent Customer                  N      Yes   
4    Financial Services  Affluent Customer                  N       No   

   tenure              address  postcode state    country  property_valuation  \
0      14     45 Shopko Center      4500   QLD  Australia                   6   
1      16    14 Mccormick Park      2113   NSW  Australia                  11   
2      10  5 Colorado Crossing      3505   VIC  Australia                   5   
3       5   207 Annamark Plaza      4814   QLD  Australia                   1   
4      19    115 Montana Place      2093   NSW  Australia                   9   

     year  month   age  
0  1957.0    7.0  64.0  
1  1970.0    3.0  51.0  
2  1974.0    8.0  47.0  
3  1979.0    1.0  42.0  
4  1965.0    9.0  56.0

In [177]:
new_customer_lists[new_customer_lists.age >= 100]

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, address, postcode, state, country, property_valuation, year, month, age]
Index: []

In [71]:
new_customer_lists['state'].value_counts()

NSW    499
VIC    258
QLD    226
Name: state, dtype: int64

In [179]:
new_customer_lists['state'] = new_customer_lists['state'].replace('QLD','Queensland')
new_customer_lists['state'] = new_customer_lists['state'].replace('NSW','New South Wales')
new_customer_lists['state'] = new_customer_lists['state'].replace('VIC', 'Victoria')
new_customer_lists.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0         Manufacturing      Mass Customer                  N      Yes   
1              Property      Mass Customer                  N       No   
2    Financial Services  Affluent Customer                  N       No   
3         Manufacturing  Affluent Customer                  N      Yes   
4    Financial Services  Affluent Customer                  N       No   

   tenure              address  postcode            state    country  \
0      14     45 Shopko Center      4500       Queensland  Australia   
1      16    14 Mccormick Park      2113  New South Wales  Australia   
2      10  5 Colorado Crossing      3505         Victoria  Australia   
3       5   207 Annamark Plaza      4814       Queensland  Australia   
4      19    115 Montana Place      2093  New South Wales  Australia   

   property_valuation    year  month   age  
0                   6  1957.0    7.0  64.0  
1                  11  1970.0    3.0  51.0  
2                   5  1974.0    8.0  47.0  
3                   1  1979.0    1.0  42.0  
4                   9  1965.0    9.0  56.0

In [76]:
new_customer_lists['state'].value_counts()

New South Wales    499
Victoria           258
Queensland         226
Name: state, dtype: int64

In [130]:
new_customer_lists['country'].value_counts()

Australia    1000
Name: country, dtype: int64

In [131]:
new_customer_lists['owns_car'].value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [132]:
new_customer_lists['job_title'].value_counts()

Associate Professor        15
Software Consultant        14
Environmental Tech         14
Chief Design Engineer      13
Assistant Manager          12
                           ..
Statistician IV             1
Web Developer II            1
Web Developer IV            1
Administrative Officer      1
Accounting Assistant II     1
Name: job_title, Length: 184, dtype: int64

In [133]:
new_customer_lists['job_industry_category'].value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [134]:
new_customer_lists['wealth_segment'].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

# 3. Exploring Customer Demographic Data Set¶


In [87]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [88]:
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [93]:
customer_demographic.drop('default', axis=1,inplace=True)
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  tenure  
0                Health      Mass Customer                  N      Yes    11.0  
1    Financial Services      Mass Customer                  N      Yes    16.0  
2              Property      Mass Customer                  N      Yes    15.0  
3                    IT      Mass Customer                  N       No     7.0  
4                   NaN  Affluent Customer                  N      Yes     8.0

In [94]:
print("Rows: ", customer_demographic.shape[0])
print("Columns: ", customer_demographic.shape[1])

Rows:  4000
Columns:  12


In [95]:
customer_demographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

In [96]:
customer_demographic.isna().values.any()

True

In [99]:
customer_demographic.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [100]:
customer_demographic.dropna(subset = ['DOB','job_title','job_industry_category','tenure'])

customer_id first_name  last_name  gender  \
0               1    Laraine  Medendorp       F   
1               2        Eli    Bockman    Male   
2               3      Arlin     Dearle    Male   
8               9       Mala       Lind  Female   
9              10   Fiorenze    Birdall  Female   
...           ...        ...        ...     ...   
3992         3993       Andi    Dumelow  Female   
3993         3994    Stephie      Byars  Female   
3994         3995      Rusty     Iapico    Male   
3995         3996    Rosalia   Halgarth  Female   
3996         3997     Blanch     Nisuis  Female   

      past_3_years_bike_related_purchases        DOB  \
0                                      93 1953-10-12   
1                                      81 1980-12-16   
2                                      61 1954-01-20   
8                                      97 1973-03-10   
9                                      49 1988-10-11   
...                                   ...        ...   
3992                                    6 1974-12-05   
3993                                    5 1989-04-07   
3994                                   93 1975-12-12   
3995                                    8 1975-08-09   
3996                                   87 2001-07-13   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
8     Business Systems Development Analyst           Argiculture   
9                  Senior Quality Engineer    Financial Services   
...                                    ...                   ...   
3992                             Librarian         Entertainment   
3993          Structural Analysis Engineer         Manufacturing   
3994                       Staff Scientist         Manufacturing   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   

         wealth_segment deceased_indicator owns_car  tenure  
0         Mass Customer                  N      Yes    11.0  
1         Mass Customer                  N      Yes    16.0  
2         Mass Customer                  N      Yes    15.0  
8     Affluent Customer                  N      Yes     8.0  
9         Mass Customer                  N      Yes    20.0  
...                 ...                ...      ...     ...  
3992      Mass Customer                  N       No    10.0  
3993  Affluent Customer                  N       No    12.0  
3994      Mass Customer                  N      Yes    14.0  
3995      Mass Customer                  N       No    19.0  
3996     High Net Worth                  N      Yes     1.0  

[2865 rows x 12 columns]

In [213]:
customer_demographic.nunique()

customer_id                            4999
first_name                             3728
last_name                              4613
gender                                    5
past_3_years_bike_related_purchases     100
DOB                                    4230
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
owns_car                                  2
tenure                                   23
address                                4989
postcode                                927
state                                     3
country                                   1
property_valuation                       12
year                                     67
month                                    12
age                                      67
age-group                                 5
age_group                                 5
dtype: int64

In [214]:
customer_demographic.gender.unique()

array(['F', 'Male', 'Female', 'Femal', 'Unspecified'], dtype=object)

In [225]:
customer_demographic['gender'] = customer_demographic['gender'].replace(['F, Femal'], 'Female')
customer_demographic['gender'] = customer_demographic['gender'].replace('M', 'Male')
customer_demographic['gender'] = customer_demographic['gender'].replace('U', 'Unspecified')

In [226]:
customer_demographic['gender'].value_counts()

Female         2550
Male           2343
Unspecified     104
Femal             1
F                 1
Name: gender, dtype: int64

In [170]:
customer_demographic['year'] = pd.DatetimeIndex(customer_demographic['DOB']).year
customer_demographic['month'] = pd.DatetimeIndex(customer_demographic['DOB']).month
customer_demographic['age'] = 2021 - customer_demographic['year']

In [160]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0                Health      Mass Customer                  N      Yes   
1    Financial Services      Mass Customer                  N      Yes   
2              Property      Mass Customer                  N      Yes   
3                    IT      Mass Customer                  N       No   
4                   NaN  Affluent Customer                  N      Yes   

   tenure              address  postcode            state    country  \
0    11.0   060 Morning Avenue    2016.0  New South Wales  Australia   
1    16.0  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2    15.0                  NaN       NaN              NaN        NaN   
3     7.0   0 Holy Cross Court    4211.0       Queensland  Australia   
4     8.0  17979 Del Mar Point    2448.0  New South Wales  Australia   

   property_valuation    year  month   age  
0                10.0  1953.0   10.0  68.0  
1                10.0  1980.0   12.0  41.0  
2                 NaN  1954.0    1.0  67.0  
3                 9.0  1961.0   10.0  60.0  
4                 4.0  1977.0    5.0  44.0

In [161]:
customer_demographic[customer_demographic.age > 100]

customer_id first_name last_name gender  \
33           34   Jephthah  Bachmann      U   

    past_3_years_bike_related_purchases        DOB        job_title  \
33                                   59 1843-12-21  Legal Assistant   

   job_industry_category     wealth_segment deceased_indicator owns_car  \
33                    IT  Affluent Customer                  N       No   

    tenure         address  postcode       state    country  \
33    20.0  833 Luster Way    4005.0  Queensland  Australia   

    property_valuation    year  month    age  
33                 8.0  1843.0   12.0  178.0

**customer_id - 34** : Age is coming **178** which is not a correct age hence will drop this value.
(Max age is 122)

In [162]:
customer_demographic.drop(33,inplace=True)

In [112]:
customer_demographic[customer_demographic.age > 100]

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, year, month, age]
Index: []

In [113]:
customer_demographic['gender'].value_counts()

Female         2039
Male           1873
Unspecified      87
Name: gender, dtype: int64

In [114]:
customer_demographic['job_title'].value_counts()

Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Administrative Assistant II              4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: job_title, Length: 195, dtype: int64

In [115]:
customer_demographic['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    222
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [116]:
customer_demographic['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     978
Name: wealth_segment, dtype: int64

In [117]:
customer_demographic['deceased_indicator'].value_counts()

N    3997
Y       2
Name: deceased_indicator, dtype: int64

In [118]:
customer_demographic['owns_car'].value_counts()

Yes    2024
No     1975
Name: owns_car, dtype: int64

# 4. Exploring Customer Address Dataset

In [120]:
customer_add.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [121]:
customer_add.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [122]:
print("Rows: ", customer_add.shape[0])
print("Columns: ", customer_add.shape[1])

Rows:  3999
Columns:  6


In [123]:
customer_add.isna().values.any()

False

In [124]:
customer_add.columns

Index(['customer_id', 'address', 'postcode', 'state', 'country',
       'property_valuation'],
      dtype='object')

In [125]:
customer_add['state'] = customer_add['state'].replace('QLD','Queensland')
customer_add['state'] = customer_add['state'].replace('NSW','New South Wales')
customer_add['state'] = customer_add['state'].replace('VIC','Victoria')

In [126]:
customer_add['state'].value_counts()

New South Wales    2140
Victoria           1021
Queensland          838
Name: state, dtype: int64

In [127]:
customer_add['postcode'].value_counts()

2170    31
2145    30
2155    30
2153    29
3977    26
        ..
3331     1
3036     1
3321     1
3305     1
2143     1
Name: postcode, Length: 873, dtype: int64

In [128]:
customer_add['country'].value_counts()

Australia    3999
Name: country, dtype: int64

In [129]:
customer_add['property_valuation'].value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

# Merging of all Customer datasets into one for analysis


All the dataset contains the customer id except NewCustomerList Dataset. So, We need to add 'Customer Id' column to new_customer_lists to enable us merge the tables vertically

In [135]:
new_customer_lists.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.57      0.7125     0.890625     0.757031   
1                  11        0.79      0.7900     0.987500     0.839375   
2                   5        0.54      0.5400     0.540000     0.540000   
3                   1        0.57      0.7125     0.712500     0.712500   
4                   9        0.86      0.8600     1.075000     1.075000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [137]:
new_customer_lists.shape

(1000, 23)

In [140]:
customer_demographic.shape

(4000, 13)

In [143]:
customer_demographic['customer_id'].iloc[-1] #returns total number of rows of customer_id columns

4000

In [144]:
#  add 'Customer Id' column to new_customer_lists
new_customer_lists.insert(0, 'customer_id', range(4001, 4001 + len(new_customer_lists)))

In [145]:
new_customer_lists.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator  ... state  \
0         Manufacturing      Mass Customer                  N  ...   QLD   
1              Property      Mass Customer                  N  ...   NSW   
2    Financial Services  Affluent Customer                  N  ...   VIC   
3         Manufacturing  Affluent Customer                  N  ...   QLD   
4    Financial Services  Affluent Customer                  N  ...   NSW   

     country property_valuation  Unnamed: 16 Unnamed: 17 Unnamed: 18  \
0  Australia                  6         0.57      0.7125    0.890625   
1  Australia                 11         0.79      0.7900    0.987500   
2  Australia                  5         0.54      0.5400    0.540000   
3  Australia                  1         0.57      0.7125    0.712500   
4  Australia                  9         0.86      0.8600    1.075000   

   Unnamed: 19  Unnamed: 20  Rank     Value  
0     0.757031            1     1  1.718750  
1     0.839375            1     1  1.718750  
2     0.540000            1     1  1.718750  
3     0.712500            4     4  1.703125  
4     1.075000            4     4  1.703125  

[5 rows x 24 columns]

In [146]:
new_customer_lists.tail()

customer_id first_name  last_name  gender  \
995         4996  Ferdinand  Romanetti    Male   
996         4997       Burk    Wortley    Male   
997         4998   Melloney      Temby  Female   
998         4999     Dickie    Cubbini    Male   
999         5000      Sylas    Duffill    Male   

     past_3_years_bike_related_purchases        DOB  \
995                                   60 1959-10-07   
996                                   22 2001-10-17   
997                                   17 1954-10-05   
998                                   30 1952-12-17   
999                                   56 1955-10-02   

                        job_title job_industry_category     wealth_segment  \
995                     Paralegal    Financial Services  Affluent Customer   
996        Senior Sales Associate                Health      Mass Customer   
997  Budget/Accounting Analyst IV    Financial Services  Affluent Customer   
998             Financial Advisor    Financial Services      Mass Customer   
999           Staff Accountant IV              Property      Mass Customer   

    deceased_indicator  ... state    country property_valuation  Unnamed: 16  \
995                  N  ...   NSW  Australia                  7         0.70   
996                  N  ...   NSW  Australia                 10         0.83   
997                  N  ...   QLD  Australia                  2         1.05   
998                  N  ...   QLD  Australia                  2         0.79   
999                  N  ...   NSW  Australia                  9         0.86   

    Unnamed: 17 Unnamed: 18  Unnamed: 19  Unnamed: 20  Rank  Value  
995      0.7000     0.70000     0.700000          996   996  0.374  
996      0.8300     1.03750     0.881875          997   997  0.357  
997      1.3125     1.31250     1.312500          997   997  0.357  
998      0.9875     0.98750     0.839375          997   997  0.357  
999      1.0750     1.34375     1.142187         1000  1000  0.340  

[5 rows x 24 columns]

#### We need to merge the Customer Demographic with the Customer Address table

In [147]:
customer_add.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211       Queensland  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216         Victoria  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [151]:
customer_add.shape

(3999, 6)

In [163]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0                Health      Mass Customer                  N      Yes   
1    Financial Services      Mass Customer                  N      Yes   
2              Property      Mass Customer                  N      Yes   
3                    IT      Mass Customer                  N       No   
4                   NaN  Affluent Customer                  N      Yes   

   tenure              address  postcode            state    country  \
0    11.0   060 Morning Avenue    2016.0  New South Wales  Australia   
1    16.0  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2    15.0                  NaN       NaN              NaN        NaN   
3     7.0   0 Holy Cross Court    4211.0       Queensland  Australia   
4     8.0  17979 Del Mar Point    2448.0  New South Wales  Australia   

   property_valuation    year  month   age  
0                10.0  1953.0   10.0  68.0  
1                10.0  1980.0   12.0  41.0  
2                 NaN  1954.0    1.0  67.0  
3                 9.0  1961.0   10.0  60.0  
4                 4.0  1977.0    5.0  44.0

In [152]:
customer_demographic.shape

(4000, 18)

In [149]:
# Merge dataframes using the customer_id column
customer_demographic = pd.merge(customer_demographic, customer_add, how='left', on='customer_id')

In [164]:
customer_demographic.tail()

customer_id first_name last_name  gender  \
3995         3996    Rosalia  Halgarth  Female   
3996         3997     Blanch    Nisuis  Female   
3997         3998     Sarene   Woolley       U   
3998         3999  Patrizius       NaN    Male   
3999         4000      Kippy   Oldland    Male   

      past_3_years_bike_related_purchases        DOB              job_title  \
3995                                    8 1975-08-09  VP Product Management   
3996                                   87 2001-07-13        Statistician II   
3997                                   60        NaT      Assistant Manager   
3998                                   11 1973-10-24                    NaN   
3999                                   76 1991-11-05   Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
3995                Health      Mass Customer                  N       No   
3996         Manufacturing     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   
3998         Manufacturing  Affluent Customer                  N      Yes   
3999                   NaN  Affluent Customer                  N       No   

      tenure                    address  postcode            state    country  \
3995    19.0         0 Transport Center    3977.0         Victoria  Australia   
3996     1.0        4 Dovetail Crossing    2350.0  New South Wales  Australia   
3997     NaN       736 Roxbury Junction    2540.0  New South Wales  Australia   
3998    10.0            1482 Hauk Trail    3064.0         Victoria  Australia   
3999    11.0  57042 Village Green Point    4511.0       Queensland  Australia   

      property_valuation    year  month   age  
3995                 6.0  1975.0    8.0  46.0  
3996                 2.0  2001.0    7.0  20.0  
3997                 6.0     NaN    NaN   NaN  
3998                 3.0  1973.0   10.0  48.0  
3999                 6.0  1991.0   11.0  30.0

In [154]:
customer_demographic.drop('default',axis=1,inplace=True)

In [165]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0                Health      Mass Customer                  N      Yes   
1    Financial Services      Mass Customer                  N      Yes   
2              Property      Mass Customer                  N      Yes   
3                    IT      Mass Customer                  N       No   
4                   NaN  Affluent Customer                  N      Yes   

   tenure              address  postcode            state    country  \
0    11.0   060 Morning Avenue    2016.0  New South Wales  Australia   
1    16.0  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2    15.0                  NaN       NaN              NaN        NaN   
3     7.0   0 Holy Cross Court    4211.0       Queensland  Australia   
4     8.0  17979 Del Mar Point    2448.0  New South Wales  Australia   

   property_valuation    year  month   age  
0                10.0  1953.0   10.0  68.0  
1                10.0  1980.0   12.0  41.0  
2                 NaN  1954.0    1.0  67.0  
3                 9.0  1961.0   10.0  60.0  
4                 4.0  1977.0    5.0  44.0

In [166]:
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3999 entries, 0 to 3999
Data columns (total 20 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3999 non-null   int64         
 1   first_name                           3999 non-null   object        
 2   last_name                            3874 non-null   object        
 3   gender                               3999 non-null   object        
 4   past_3_years_bike_related_purchases  3999 non-null   int64         
 5   DOB                                  3912 non-null   datetime64[ns]
 6   job_title                            3493 non-null   object        
 7   job_industry_category                3343 non-null   object        
 8   wealth_segment                       3999 non-null   object        
 9   deceased_indicator                   3999 non-null   object        
 10  owns_car    

In [171]:
customer_demographic['age-group'] = pd.cut(customer_demographic['age'], [0,30,40,50,60,200], labels = ['<30','30-40','40-50','50-60','60+'])

In [172]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  ... tenure  \
0                Health      Mass Customer                  N  ...   11.0   
1    Financial Services      Mass Customer                  N  ...   16.0   
2              Property      Mass Customer                  N  ...   15.0   
3                    IT      Mass Customer                  N  ...    7.0   
4                   NaN  Affluent Customer                  N  ...    8.0   

               address postcode            state    country  \
0   060 Morning Avenue   2016.0  New South Wales  Australia   
1  6 Meadow Vale Court   2153.0  New South Wales  Australia   
2                  NaN      NaN              NaN        NaN   
3   0 Holy Cross Court   4211.0       Queensland  Australia   
4  17979 Del Mar Point   2448.0  New South Wales  Australia   

  property_valuation    year  month   age  age-group  
0               10.0  1953.0   10.0  68.0        60+  
1               10.0  1980.0   12.0  41.0      40-50  
2                NaN  1954.0    1.0  67.0        60+  
3                9.0  1961.0   10.0  60.0      50-60  
4                4.0  1977.0    5.0  44.0      40-50  

[5 rows x 21 columns]

In [180]:
new_customer_lists.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0         Manufacturing      Mass Customer                  N      Yes   
1              Property      Mass Customer                  N       No   
2    Financial Services  Affluent Customer                  N       No   
3         Manufacturing  Affluent Customer                  N      Yes   
4    Financial Services  Affluent Customer                  N       No   

   tenure              address  postcode            state    country  \
0      14     45 Shopko Center      4500       Queensland  Australia   
1      16    14 Mccormick Park      2113  New South Wales  Australia   
2      10  5 Colorado Crossing      3505         Victoria  Australia   
3       5   207 Annamark Plaza      4814       Queensland  Australia   
4      19    115 Montana Place      2093  New South Wales  Australia   

   property_valuation    year  month   age  
0                   6  1957.0    7.0  64.0  
1                  11  1970.0    3.0  51.0  
2                   5  1974.0    8.0  47.0  
3                   1  1979.0    1.0  42.0  
4                   9  1965.0    9.0  56.0

In [183]:
new_customer_lists["age_group"] = pd.cut(new_customer_lists.age, [0,30,40,50,60,9999], labels=['<30','30-40','40-50','50-60','60+'])

In [184]:
new_customer_lists.head()

customer_id first_name  last_name  gender  \
0         4001    Chickie    Brister    Male   
1         4002      Morly     Genery    Male   
2         4003    Ardelis  Forrester  Female   
3         4004     Lucine      Stutt  Female   
4         4005    Melinda     Hadlee  Female   

   past_3_years_bike_related_purchases        DOB                   job_title  \
0                                   86 1957-07-12             General Manager   
1                                   69 1970-03-22         Structural Engineer   
2                                   10 1974-08-28      Senior Cost Accountant   
3                                   64 1979-01-28  Account Representative III   
4                                   34 1965-09-21           Financial Analyst   

  job_industry_category     wealth_segment deceased_indicator  ... tenure  \
0         Manufacturing      Mass Customer                  N  ...     14   
1              Property      Mass Customer                  N  ...     16   
2    Financial Services  Affluent Customer                  N  ...     10   
3         Manufacturing  Affluent Customer                  N  ...      5   
4    Financial Services  Affluent Customer                  N  ...     19   

               address postcode            state    country  \
0     45 Shopko Center     4500       Queensland  Australia   
1    14 Mccormick Park     2113  New South Wales  Australia   
2  5 Colorado Crossing     3505         Victoria  Australia   
3   207 Annamark Plaza     4814       Queensland  Australia   
4    115 Montana Place     2093  New South Wales  Australia   

  property_valuation    year  month   age  age_group  
0                  6  1957.0    7.0  64.0        60+  
1                 11  1970.0    3.0  51.0      50-60  
2                  5  1974.0    8.0  47.0      40-50  
3                  1  1979.0    1.0  42.0      40-50  
4                  9  1965.0    9.0  56.0      50-60  

[5 rows x 21 columns]

In [185]:
new_customer_lists.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          1000 non-null   int64         
 1   first_name                           1000 non-null   object        
 2   last_name                            971 non-null    object        
 3   gender                               1000 non-null   object        
 4   past_3_years_bike_related_purchases  1000 non-null   int64         
 5   DOB                                  983 non-null    datetime64[ns]
 6   job_title                            894 non-null    object        
 7   job_industry_category                835 non-null    object        
 8   wealth_segment                       1000 non-null   object        
 9   deceased_indicator                   1000 non-null   object        
 10  owns_car     

In [186]:
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3999 entries, 0 to 3999
Data columns (total 21 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3999 non-null   int64         
 1   first_name                           3999 non-null   object        
 2   last_name                            3874 non-null   object        
 3   gender                               3999 non-null   object        
 4   past_3_years_bike_related_purchases  3999 non-null   int64         
 5   DOB                                  3912 non-null   datetime64[ns]
 6   job_title                            3493 non-null   object        
 7   job_industry_category                3343 non-null   object        
 8   wealth_segment                       3999 non-null   object        
 9   deceased_indicator                   3999 non-null   object        
 10  owns_car    

From the above information both dataframe have equal columns so 
will merge both data sets: **new_customer_lists & customer_demographic**

In [187]:
df = pd.concat([customer_demographic, new_customer_lists], ignore_index=True, sort=False)

In [229]:
customer_demographic = df

In [231]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  ...  \
0                Health      Mass Customer                  N  ...   
1    Financial Services      Mass Customer                  N  ...   
2              Property      Mass Customer                  N  ...   
3                    IT      Mass Customer                  N  ...   
4                   NaN  Affluent Customer                  N  ...   

               address  postcode            state    country  \
0   060 Morning Avenue    2016.0  New South Wales  Australia   
1  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2                  NaN       NaN              NaN        NaN   
3   0 Holy Cross Court    4211.0       Queensland  Australia   
4  17979 Del Mar Point    2448.0  New South Wales  Australia   

  property_valuation    year  month   age  age-group  age_group  
0               10.0  1953.0   10.0  68.0        60+        NaN  
1               10.0  1980.0   12.0  41.0      40-50        NaN  
2                NaN  1954.0    1.0  67.0        60+        NaN  
3                9.0  1961.0   10.0  60.0      50-60        NaN  
4                4.0  1977.0    5.0  44.0      40-50        NaN  

[5 rows x 22 columns]

#### Merge Transction dataframe with customer demographic dataframe to make a one single dataframe 

In [198]:
transactions_customer = transactions

In [199]:
transactions_customer.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date   profit  
0       71.49          53.62                  41245.0    17.87  
1     2091.47         388.92                  41701.0  1702.55  
2     1793.43         248.82                  36361.0  1544.61  
3     1198.46         381.10                  36145.0   817.36  
4     1765.30         709.48                  42226.0  1055.82

In [244]:
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","gender"]], on="customer_id", how="left")

In [245]:
transactions_customer.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date   profit  gender  
0       71.49          53.62                  41245.0    17.87    Male  
1     2091.47         388.92                  41701.0  1702.55  Female  
2     1793.43         248.82                  36361.0  1544.61    Male  
3     1198.46         381.10                  36145.0   817.36    Male  
4     1765.30         709.48                  42226.0  1055.82  Female

In [246]:
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","past_3_years_bike_related_purchases"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","age"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","age_group"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","job_title"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","job_industry_category"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","wealth_segment"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","owns_car"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","postcode"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","state"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","property_valuation"]], on="customer_id", how="left")
transactions_customer = pd.merge(transactions_customer,customer_demographic[["customer_id","country"]], on="customer_id", how="left")


In [247]:
transactions_customer.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  ...  \
0     Approved           Solex     Standard        medium       medium  ...   
1     Approved   Trek Bicycles     Standard        medium        large  ...   
2     Approved      OHM Cycles     Standard           low       medium  ...   
3     Approved  Norco Bicycles     Standard        medium       medium  ...   
4     Approved  Giant Bicycles     Standard        medium        large  ...   

    age  age_group                   job_title  job_industry_category  \
0  66.0        NaN         Software Engineer I     Financial Services   
1  42.0        NaN         Clinical Specialist                 Health   
2  44.0        NaN  Desktop Support Technician                 Retail   
3  59.0        NaN             Staff Scientist     Financial Services   
4  25.0        NaN                         NaN                 Retail   

      wealth_segment  owns_car  postcode            state property_valuation  \
0      Mass Customer       Yes    3064.0         Victoria                6.0   
1      Mass Customer       Yes    2196.0  New South Wales                5.0   
2  Affluent Customer        No    2835.0  New South Wales                1.0   
3      Mass Customer        No    2096.0  New South Wales               10.0   
4  Affluent Customer       Yes    2292.0  New South Wales                7.0   

     country  
0  Australia  
1  Australia  
2  Australia  
3  Australia  
4  Australia  

[5 rows x 26 columns]

In [248]:
transactions_customer.sort_values(["transaction_date"], ascending=False)

transaction_id  product_id  customer_id transaction_date  online_order  \
1696             1697          56         2421       2017-12-30           1.0   
15269           15270           3          195       2017-12-30           0.0   
4668             4669          29         1052       2017-12-30           0.0   
13574           13575          98         2396       2017-12-30           0.0   
3937             3938          95          993       2017-12-30           0.0   
...               ...         ...          ...              ...           ...   
16397           16398          43         2160       2017-01-01           1.0   
15204           15205          87          123       2017-01-01           0.0   
15137           15138           3         1022       2017-01-01           1.0   
3459             3460          64          234       2017-01-01           1.0   
13933           13934          94          824       2017-01-01           0.0   

      order_status           brand product_line product_class product_size  \
1696      Approved      OHM Cycles     Standard        medium       medium   
15269     Approved   Trek Bicycles     Standard        medium        large   
4668      Approved        WeareA2B     Standard        medium       medium   
13574     Approved   Trek Bicycles     Standard          high       medium   
3937      Approved      OHM Cycles      Touring           low       medium   
...            ...             ...          ...           ...          ...   
16397     Approved           Solex     Standard        medium       medium   
15204     Approved      OHM Cycles     Standard        medium       medium   
15137     Approved   Trek Bicycles     Standard        medium        large   
3459      Approved  Giant Bicycles     Standard          high        small   
13933     Approved  Giant Bicycles     Standard        medium        large   

       ...   age  age_group              job_title  job_industry_category  \
1696   ...  67.0        NaN         Research Nurse                 Health   
15269  ...  63.0        NaN        Project Manager          Manufacturing   
4668   ...  35.0        NaN       Junior Executive     Financial Services   
13574  ...  48.0        NaN              Recruiter                 Retail   
3937   ...  44.0        NaN    Account Coordinator            Argiculture   
...    ...   ...        ...                    ...                    ...   
16397  ...  50.0        NaN              Recruiter                 Health   
15204  ...  33.0        NaN     Analyst Programmer     Financial Services   
15137  ...  31.0        NaN      Assistant Manager     Financial Services   
3459   ...  52.0        NaN  Chief Design Engineer                 Retail   
13933  ...  44.0        NaN      Financial Advisor     Financial Services   

          wealth_segment  owns_car  postcode            state  \
1696   Affluent Customer       Yes    3046.0         Victoria   
15269      Mass Customer        No    3143.0         Victoria   
4668      High Net Worth        No    3184.0         Victoria   
13574  Affluent Customer       Yes    4128.0       Queensland   
3937      High Net Worth        No    2871.0  New South Wales   
...                  ...       ...       ...              ...   
16397      Mass Customer       Yes    3172.0         Victoria   
15204     High Net Worth        No    2516.0  New South Wales   
15137  Affluent Customer       Yes    2136.0  New South Wales   
3459      High Net Worth        No    2036.0  New South Wales   
13933  Affluent Customer       Yes    4127.0       Queensland   

      property_valuation    country  
1696                 8.0  Australia  
15269                9.0  Australia  
4668                11.0  Australia  
13574                7.0  Australia  
3937                 3.0  Australia  
...                  ...        ...  
16397                8.0  Australia  
15204                8.0  Australia  
15137               10.0  Australia  
3459       

In [255]:
transactions_customer.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  ...  \
0     Approved           Solex     Standard        medium       medium  ...   
1     Approved   Trek Bicycles     Standard        medium        large  ...   
2     Approved      OHM Cycles     Standard           low       medium  ...   
3     Approved  Norco Bicycles     Standard        medium       medium  ...   
4     Approved  Giant Bicycles     Standard        medium        large  ...   

    age  age_group                   job_title  job_industry_category  \
0  66.0        NaN         Software Engineer I     Financial Services   
1  42.0        NaN         Clinical Specialist                 Health   
2  44.0        NaN  Desktop Support Technician                 Retail   
3  59.0        NaN             Staff Scientist     Financial Services   
4  25.0        NaN                         NaN                 Retail   

      wealth_segment  owns_car  postcode            state property_valuation  \
0      Mass Customer       Yes    3064.0         Victoria                6.0   
1      Mass Customer       Yes    2196.0  New South Wales                5.0   
2  Affluent Customer        No    2835.0  New South Wales                1.0   
3      Mass Customer        No    2096.0  New South Wales               10.0   
4  Affluent Customer       Yes    2292.0  New South Wales                7.0   

     country  
0  Australia  
1  Australia  
2  Australia  
3  Australia  
4  Australia  

[5 rows x 26 columns]